File: react-native/src/interpolate.test.ts
// @ts-nocheck — no test-runner types installed; run with jest or vitest
import { buildVariableContext, interpolate, normalizeVariable } from './interpolate';
import vectors from './expr-test-vectors.json';

type VarEntry = { type: 'string' | 'number'; value?: string };
type FallbackMap = Record<string, string>;
type VarsMap = Record<string, VarEntry | FallbackMap>;

interface TestVector {
    description: string;
    expr: string;
    vars: VarsMap;
    out: string;
}

describe('interpolate — expression test vectors', () => {
    (vectors as TestVector[]).forEach((vector) => {
        it(vector.description, () => {
            const { vars, expr, out } = vector;

            // Extract _fallback map (if present)
            const fallbackEntry = vars['_fallback'] as FallbackMap | undefined;

            // Build schemas and cepVars, skipping the _fallback meta-key
            const schemas = Object.entries(vars)
                .filter(([key]) => key !== '_fallback')
                .map(([name, entry]) => {
                    const varEntry = entry as VarEntry;
                    return normalizeVariable({
                        name,
                        type: varEntry.type,
                        fallbackValue: fallbackEntry?.[name] ?? '',
                    });
                });

            const cepVars: Record<string, string> = {};
            for (const [name, entry] of Object.entries(vars)) {
                if (name === '_fallback') continue;
                const varEntry = entry as VarEntry;
                if (varEntry.value !== undefined) {
                    cepVars[name] = varEntry.value;
                }
            }

            const context = buildVariableContext(schemas, cepVars);
            const result = interpolate(`{{ ${expr} }}`, context);

            expect(result).toBe(out);
        });
    });
});

File: android/digia-engage/src/test/java/com/digia/engage/internal/InterpolateTest.kt
package com.digia.engage.internal

import org.json.JSONArray
import org.junit.Assert.assertEquals
import org.junit.Test
import java.io.File

class InterpolateTest {
    @Test
    fun goldenVectors() {
        val json = File("src/test/resources/expr-test-vectors.json").readText()
        val vectors = JSONArray(json)
        for (i in 0 until vectors.length()) {
            val v = vectors.getJSONObject(i)
            val expr = v.getString("expr")
            val out = v.getString("out")
            val vars = v.getJSONObject("vars")
            val fallbackMap = if (vars.has("_fallback")) {
                val fb = vars.getJSONObject("_fallback")
                fb.keys().asSequence().associateWith { fb.getString(it) }
            } else emptyMap()
            val schemas = mutableListOf<VariableSchema>()
            val cepVars = mutableMapOf<String, String>()
            for (name in vars.keys()) {
                if (name == "_fallback") continue
                val entry = vars.getJSONObject(name)
                val type = entry.optString("type", "string")
                val fb = fallbackMap[name] ?: ""
                schemas.add(VariableSchema(name, type, fb))
                if (entry.has("value")) cepVars[name] = entry.getString("value")
            }
            val context = buildVariableContext(schemas, cepVars)
            val result = interpolate("{{ $expr }}", context)
            assertEquals("vector[$i] '${v.optString("description", expr)}'", out, result)
        }
    }
}
